src/OpenRiaServices.DomainServices.Hosting/Test/Data/BufferManagerStreamTests.cs

```diff
@@ -0,0 +1,201 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.ServiceModel.Channels;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using OpenRiaServices.DomainServices.Client.Test;
+using BufferManagerStream = OpenRiaServices.DomainServices.Hosting.PoxBinaryMessageEncodingBindingElement.PoxBinaryMessageEncoder.BufferManagerStream;
+
+namespace OpenRiaServices.DomainServices.Hosting.Test.Data
+{
+    [TestClass]
+    public class BufferManagerStreamTests
+    {
+        // Buffer with bytes [0..255]
+        private readonly byte[] _input;
+
+        public BufferManagerStreamTests()
+        {
+            _input = new byte[256];
+            for (int i = 0; i < _input.Length; ++i)
+            {
+                _input[i] = (byte)i;
+            }
+        }
+
+        /// <summary>
+        /// Only partially fill first buffer
+        /// </summary>
+        [TestMethod]
+        public void SingleSmallWrite()
+        {
+            var manager = new BufferManageMock();
+            using (var stream = new BufferManagerStream(manager, 0, 4, 1024))
+            {
+                stream.Write(_input, 0, 2);
+
+                var buffer = VerifyStreamContents(stream, 0, 2, manager);
+                Assert.AreEqual(1, manager.Allocated.Count, "Should only have allocated a single buffer");
+                Assert.AreSame(manager.Allocated[0], buffer.Array, "Should reuse initial array");
+            }
+        }
+
+        /// <summary>
+        /// Only partially fill first buffer
+        /// </summary>
+        [TestMethod]
+        public void FillInitialBuffer()
+        {
+            int initialOffset = 0, minBufferSize = 4;
```

Comment:
I am unsure if it makes sense to add tests with initialOffset != 0 it does not happen for GET request but maybe for POST